File: java/it/eja/surf/Javascript.java
package it.eja.surf;

import android.webkit.JavascriptInterface;

import org.json.JSONException;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLConnection;

public class Javascript {

    private boolean authCheck() {
        return Setting.host.equals(MainActivity.hostCurrent);
    }

    @JavascriptInterface
    public String shell(String cmd) {
        if (authCheck()) {
            try {
                Process process = Runtime.getRuntime().exec(cmd);
                BufferedReader reader = new BufferedReader(
                        new InputStreamReader(process.getInputStream())
                );
                char[] buffer = new char[4096];
                StringBuilder output = new StringBuilder();
                int read;
                while ((read = reader.read(buffer)) > 0) {
                    output.append(buffer, 0, read);
                }
                reader.close();
                process.waitFor();
                return output.toString();
            } catch (IOException | InterruptedException e) {
                throw new RuntimeException(e);
            }
        } else {
            return "";
        }
    }

    @JavascriptInterface
    public void fileDownload(String url, String outputFile) {
        if (authCheck()) {
            try {
                URL u = new URL(url);
                URLConnection conn = u.openConnection();
                int contentLength = conn.getContentLength();
                DataInputStream stream = new DataInputStream(u.openStream());
                byte[] buffer = new byte[contentLength];
                stream.readFully(buffer);
                stream.close();
                DataOutputStream fos = new DataOutputStream(new FileOutputStream(Setting.path + File.separator + outputFile));
                fos.write(buffer);
                fos.flush();
                fos.close();
            } catch (IOException ignored) {
            }
        }
    }

    @JavascriptInterface
    public String fileRead(String fileName) {
        if (authCheck()) {
            return Setting.fileRead(fileName);
        } else {
            return "";
        }
    }

    @JavascriptInterface
    public void fileWrite(String fileName, String value) {
        if (authCheck()) {
            Setting.fileWrite(fileName, value);
        }
    }

    @JavascriptInterface
    public String settingGetAll() throws JSONException {
        if (authCheck()) {
            return Setting.load();
        } else {
            return "";
        }
    }

    @JavascriptInterface
    public String settingGet(String name) throws JSONException {
        if (authCheck()) {
            if (name.isEmpty()) {
                return Setting.load();
            } else {
                return Setting.eja.getString(name);
            }
        } else {
            return "";
        }
    }

    @JavascriptInterface
    public void settingPut(String name, String value) throws JSONException {
        if (authCheck()) {
            Setting.eja.put(name, value);
            Setting.save();
        }
    }

    @JavascriptInterface
    public void settingInit() throws JSONException {
        if (authCheck()) {
            Setting.load();
        }
    }

    @JavascriptInterface
    public void bookAdd(String value) {
        if (authCheck()) {
            Setting.bookAdd(value);
        }
    }

    @JavascriptInterface
    public String bookRead() {
        if (authCheck()) {
            return Setting.fileRead("eja.book");
        } else {
            return "";
        }
    }

    @JavascriptInterface
    public void bookRemove(String index) {
        if (authCheck()) {
            Setting.bookRemove(Integer.parseInt(index));
        }
    }
}

File: app/src/main/java/it/eja/surf/Javascript.kt
// Copyright (C) 2021-2024 by Ubaldo Porcheddu <ubaldo@eja.it>

package it.eja.surf

import android.webkit.JavascriptInterface
import org.json.JSONException
import java.io.*
import java.net.URL

class Javascript {

    private fun authCheck(): Boolean {
        return Setting.host == MainActivity.hostCurrent
    }

    @JavascriptInterface
    fun shell(cmd: String?): String {
        if (!authCheck() || cmd == null) return ""
        return try {
            val process = Runtime.getRuntime().exec(cmd)
            val output = StringBuilder()
            BufferedReader(InputStreamReader(process.inputStream)).use { reader ->
                val buffer = CharArray(4096)
                var read: Int
                while (reader.read(buffer).also { read = it } > 0) {
                    output.append(buffer, 0, read)
                }
            }
            process.waitFor()
            output.toString()
        } catch (e: IOException) {
            ""
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
            ""
        }
    }

    @JavascriptInterface
    fun fileDownload(url: String?, outputFile: String) {
        if (!authCheck() || url.isNullOrEmpty()) return
        try {
            val u = URL(url)
            val conn = u.openConnection()
            val contentLength = conn.contentLength
            DataInputStream(u.openStream()).use { stream ->
                val buffer = ByteArray(contentLength)
                stream.readFully(buffer)
                DataOutputStream(FileOutputStream("${Setting.path}${File.separator}$outputFile")).use { fos ->
                    fos.write(buffer)
                    fos.flush()
                }
            }
        } catch (ignored: IOException) {
        }
    }

    @JavascriptInterface
    fun fileRead(fileName: String?): String {
        return if (authCheck()) {
            Setting.fileRead(fileName)
        } else {
            ""
        }
    }

    @JavascriptInterface
    fun fileWrite(fileName: String?, value: String?) {
        if (authCheck()) {
            Setting.fileWrite(fileName, value)
        }
    }

    @JavascriptInterface
    @Throws(JSONException::class)
    fun settingGetAll(): String {
        return if (authCheck()) {
            Setting.load()
        } else {
            ""
        }
    }

    @JavascriptInterface
    @Throws(JSONException::class)
    fun settingGet(name: String): String? {
        return if (authCheck()) {
            if (name.isEmpty()) {
                Setting.load()
            } else {
                Setting.eja.getString(name)
            }
        } else {
            ""
        }
    }

    @JavascriptInterface
    @Throws(JSONException::class)
    fun settingPut(name: String?, value: String?) {
        if (authCheck()) {
            Setting.eja.put(name, value)
            Setting.save()
        }
    }

    @JavascriptInterface
    @Throws(JSONException::class)
    fun settingInit() {
        if (authCheck()) {
            Setting.load()
        }
    }

    @JavascriptInterface
    fun bookAdd(value: String) {
        if (authCheck()) {
            Setting.bookAdd(value)
        }
    }

    @JavascriptInterface
    fun bookRead(): String {
        return if (authCheck()) {
            Setting.fileRead("eja.book")
        } else {
            ""
        }
    }

    @JavascriptInterface
    fun bookRemove(index: String) {
        if (authCheck()) {
            try {
                Setting.bookRemove(index.toInt())
            } catch (e: NumberFormatException) {
                // Handle the error, maybe log it
            }
        }
    }
}